types/pouchdb-mapreduce/index.d.ts

```diff
@@ -91,17 +94,17 @@ declare namespace PouchDB {
          * Invoke a map/reduce function, which allows you to perform more complex queries
          * on PouchDB than what you get with allDocs().
          */
-        query(fun: string | Filter | Function, opts: Query.Options, callback: (err: Core.Error, res: Query.Response<Content>) => void): void;
+        query<Model>(fun: string | Filter | ((doc: any) => void), opts: Query.Options, callback: (err: Core.Error, res: Query.Response<Content & Model>) => void): void;
         /**
          * Invoke a map/reduce function, which allows you to perform more complex queries
          * on PouchDB than what you get with allDocs().
          */
-        query(fun: string | Filter | Function, callback: (err: Core.Error, res: Query.Response<Content>) => void): void;
+        query<Model>(fun: string | Filter | ((doc: any) => void), callback: (err: Core.Error, res: Query.Response<Content & Model>) => void): void;
         /**
          * Invoke a map/reduce function, which allows you to perform more complex queries
          * on PouchDB than what you get with allDocs().
          */
-        query(fun: string | Filter | Function, opts?: Query.Options): Promise<Query.Response<Content>>;
+        query<Model>(fun: string | Filter | ((doc: any) => void), opts?: Query.Options): Promise<Query.Response<Content & Model>>;
```

Comment:
But on the other side, on a heterogeneous DB we'll have the filter function executed on top of any concrete type we might have there, which, on a trully heterogeneous DB, will be many and diverse. If we type `doc` to `Content & Model` we'll be narrowing past the runtime behaviour. I'd say that `doc: Content` is safe everytime though, which is miles better than `any`.